cpp/src/allgather.cpp

```diff
@@ -0,0 +1,149 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <algorithm>
+#include <numeric>
+#include <unordered_map>
+#include <vector>
+
+#include <rapidsmpf/allgather/allgather.hpp>
+#include <rapidsmpf/buffer/packed_data.hpp>
+#include <rapidsmpf/buffer/resource.hpp>
+#include <rapidsmpf/communicator/communicator.hpp>
+#include <rapidsmpf/progress_thread.hpp>
+#include <rapidsmpf/shuffler/chunk.hpp>
+#include <rapidsmpf/shuffler/shuffler.hpp>
+#include <rapidsmpf/statistics.hpp>
+
+namespace rapidsmpf::allgather {
+
+/**
+ * @brief A naive implementation of the all-gather operation.
+ *
+ * The implementation creates a copy of the data for each peer during insertion and assign
+ * the peer's rank as the partition ID, i.e., there will only be one partition per rank.
+ * `partition_owner` lambda function is an identity function. These enable the data
+ * copies to be sent to the corresponding rank using the Shuffler.
+ *
+ * @note This implementation might not be scalable, as it makes a copy of the data for
+ * each peer during insertion.
+ */
+
+AllGather::AllGather(
+    std::shared_ptr<Communicator> comm,
+    std::shared_ptr<ProgressThread> progress_thread,
+    OpID op_id,
+    rmm::cuda_stream_view stream,
+    BufferResource* br,
+    std::shared_ptr<Statistics> statistics
+)
+    : comm_(comm.get()),
+      shuffler_(
+          std::make_unique<shuffler::Shuffler>(
+              std::move(comm),
+              std::move(progress_thread),
+              op_id,
+              comm_->nranks(),
+              stream,
+              br,
+              std::move(statistics),
+              [](std::shared_ptr<Communicator> const&, shuffler::PartID pid) -> Rank {
+                  // identity-like mapping, as there are only n_ranks partitions
+                  return static_cast<Rank>(pid);
+              }
+          )
+      ),
+      stream_(stream),
+      br_(br) {}
+
+AllGather::~AllGather() {
+    shutdown();
+}
+
+void AllGather::shutdown() {
+    if (shuffler_) {
+        shuffler_->shutdown();
+        shuffler_.reset();
+    }
```

Comment:
Surely just `shuffler_.reset()` since the dtor for the shuffler will run `shutdown()` for us.